### PR TITLE
WIP: Composer Jenkinsfile

### PR DESCRIPTION
**Pull request by ravi-mosaicml**
This PR removes github actions and switches to having tests run on Jenkins:

TODO:

- [x] @nlsapp @connor-m-kaz Configure the new jenkins cluster. Ensure that the throttling is set correctly, as we are running into websocket timeout issues when the parallel builds scale to much
- [ ] ~~@nlsapp @connor-m-kaz Configure the [remote jenkinsfile plugin](https://plugins.jenkins.io/remote-file/) so we can store the `.ci/Jenkinsfile` in a separate repo (such as https://github.com/mosaicml/testing)~~ Deferring for #247
- [ ] ~~@ravi-mosaicml Write a stub jenkinsfile that the above remote plugin would point to. This stub file should load the correct jenkinsfile for the repo~~ Deferring for #247
- [ ] ~~@nlsapp @connor-m-kaz Configure jenkins reports to be public (without authentication), so public users can view Jenkins test reports which are auto-linked to the PR~~ Deferring for #248
- [ ] ~~@ravi-mosaicml Configure commit trigger strings that can run specific jenkins jobs~~ Right now, let's run everything.
- [x] Merge #182
- [x] Merge #184 
- [x] Configure jenkins to run deepspeed tests separately from gpu tests (see the new marker in #199)
- [x] Merge #185 
- [x] Merge #188 
- [x] Merge #215
- [x] Merge #233

### CODE REVIEW

File: .ci/Jenkinsfile [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+                string(name: 'P_N_GPUS', value: nGpus),
+            ]
+        )
+    }
+}
+
+stage('Prepare') {
+    node (pCloud) {
+        // In order to print the env and get the commit url and hash, we must be on a real agent
+        sh 'printenv'
```
Comment (nlsapp) [inline .ci/Jenkinsfile:75]: Debug?

Printing the env might leak secrets, if they are mounted as env vars (like the docker secret for kaniko)

File: .ci/Jenkinsfile [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+
+        def loadedSCM = checkout scm
+
+        sh 'printenv'
```
Comment (nlsapp) [inline .ci/Jenkinsfile:79]: Debug?

File: .ci/Jenkinsfile [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+            P_CLOUD: pCloud,
+            P_BUILD_MATRIX: './composer/pytorch_build_matrix.sh',
+            P_BUILD_MATRIX_GIT_REPO: 'https://github.com/mosaicml/testing.git',  // TODO RJPP_SCM_URL
+            P_BUILD_MATRIX_GIT_COMMIT: 'main', // TODO RJPP_BRANCH
+            P_DOCKERFILE: 'Dockerfile',
+            P_BUILD_CONTEXT: './docker/pytorch',
+            P_GIT_REPO: gitUrl,
+            P_GIT_COMMIT: gitCommit,
+            P_CPU_LIMIT: '4',
+            P_MEM_LIMIT: '15Gi',
```
Comment (nlsapp) [inline .ci/Jenkinsfile:138]: Do we really need this amount of resources for the docker build? 4 cores and 15Gi of ram seems high

File: .ci/Jenkinsfile [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+        builds << build(
+            job: 'scratch/pytest',
+            parameters: [
+                string(name: 'P_CLOUD', value: pCloud),
+                string(name: 'P_GIT_REPO', value: gitUrl),
+                string(name: 'P_GIT_COMMIT', value: gitCommit),
+                string(name: 'P_DOCKER_IMAGE', value: pDockerImage),
+                string(name: 'P_CPU_LIMIT', value: cpuLimit),
+                string(name: 'P_MEM_LIMIT', value: memLimit),
+                string(name: 'P_TIMEOUT', value: '1800'),
```
Comment (nlsapp) [inline .ci/Jenkinsfile:62]: TODO - move timeout values to global vars (or something), to make updates easier

File: .ci/Jenkinsfile [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+        pytestCliArgs = "$pytestCliArgs -m $marker"
+    }
+    runWithChecks(
+        name: name,
+        title: title,
+        summary: summary,
+    ) {
+        echo "$title"
+        builds << build(
+            job: 'scratch/pytest',
```
Comment (nlsapp) [inline .ci/Jenkinsfile:54]: TODO - have the first part of this job path be a var - so we can easily move it in the future

maybe something like: `${jenkinsJobBasePath}/pytest`

File: .ci/Jenkinsfile [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+                    title: 'Lint',
+                    summary: 'Static Analysis Checks',
+                ) {
+                    builds << build(
+                        job: 'scratch/lint',
+                        parameters: [
+                            string(name: 'P_CLOUD', value: pCloud),
+                            string(name: 'P_GIT_REPO', value: gitUrl),
+                            string(name: 'P_GIT_COMMIT', value: gitCommit),
+                            string(name: 'P_DOCKER_IMAGE', value: "mosaicml/pytorch:1.10.0_cpu-python3.9-ubuntu20.04"),
```
Comment (nlsapp) [inline .ci/Jenkinsfile:158]: shouldn't this image be calculated like it is above?

`"mosaicml/pytorch:${pytorchVersion}_${cudaVersion}-python${pythonVersion}-ubuntu20.04"`

File: .ci/Jenkinsfile [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+            'Python 3.9 - Dev': { -> runPytest("3.9", "dev") },
+            'Python 3.9 - All (GPU)': { -> runPytest("3.9", "all", true, "gpu") },
+            'Python 3.9 - All (GPU-Deepspeed)': { -> runPytest("3.9", "all", true, "deepspeed") },
+        ]        
+    }
+    try {
+        parallel(jobs)
+    }
+    finally {
+        stage('Merge Artifcats') {
```
Comment (nlsapp) [inline .ci/Jenkinsfile:179]: `Artifacts`

😸

File: .ci/Jenkinsfile [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+                        optional: true,
+                    )
+                }
+
+                sh 'ls -al'
+
+                sh 'mkdir -p build/output/'
+
+                sh 'ls -al build/output/'
+                archiveArtifacts(artifacts: "build/output/*.xml", fingerprint: true, allowEmptyArchive: true)
```
Comment (nlsapp) [inline .ci/Jenkinsfile:196]: This is fine, for now, but Jenkins will not serve as our long term artifact repo - and there's no guarantee these will be retained

We need to either push things we want to keep to GH, S3, or we look at setting something like artifactory up

File: docker/Makefile [hunk 9]
```diff
@@ -1,5 +1,5 @@
 # Base image to use for the docker build.
-BASE_IMAGE ?= 'nvidia/cuda:11.3.1-cudnn8-runtime-ubuntu20.04'
+BASE_IMAGE ?= 'nvidia/cuda:11.3.1-cudnn8-devel-ubuntu20.04'
```
Comment (nlsapp) [inline docker/Makefile:2]: We really need devel? 😢  that makes our images much larger

**Review by nlsapp — CHANGES_REQUESTED** [review]
Cleanup debug statements, check the questions, otherwise looks good

File: .ci/Jenkinsfile (hunk 3 above)
Comment (ravi-mosaicml) [reply .ci/Jenkinsfile:138]: I was getting sigkill errors on the build. We do have plenty of extra resources on the build agent so would prefer to leave as is.

File: .ci/Jenkinsfile (hunk 6 above)
Comment (ravi-mosaicml) [reply .ci/Jenkinsfile:158]: We only run linting on the latest version.

File: docker/Makefile (hunk 9 above)
Comment (ravi-mosaicml) [reply docker/Makefile:2]: Yup, unfortunately we need it for A100s on GCP. See https://github.com/mosaicml/composer/blob/bf0bd3b3fa0bab7f3dd06cf7a9b1d55e7a8cc31b/docker/pytorch/Dockerfile#L36

File: .ci/Jenkinsfile (hunk 8 above)
Comment (ravi-mosaicml) [reply .ci/Jenkinsfile:196]: Right now the only artifacts are coverage reports -- which I'm ok with losing. But yes, once we start doing real test regressions, we'll need to set up a real artifact repo.

**Review by nlsapp — APPROVED** [review]
LGTM, although I'd prefer to see the commit counter hit 200 🤣

File: .ci/Jenkinsfile [hunk 10]
```diff
@@ -25,35 +25,85 @@ def cloneJenkinsfilesRepo() {
...
     }
-    runWithChecks(
-        name: name,
-        title: title,
-        summary: summary,
-    ) {
-        echo "$title"
+    def summary = title
+
+    def pytestCommand = """#!/bin/bash
```
Comment (nlsapp) [inline .ci/Jenkinsfile:69]: Safer Shebangs: `#!/usr/bin/env bash`

File: scripts/lint.sh [hunk 11]
```diff
@@ -0,0 +1,29 @@
+#!/bin/bash
```
Comment (nlsapp) [inline scripts/lint.sh:1]: Safer Shebangs: `#!/usr/bin/env bash`

**Comment by ravi-mosaicml** [comment]
> LGTM, although I'd prefer to see the commit counter hit 200 rofl

Feedback addressed :rofl:

**Review by nlsapp — APPROVED** [review]
Approved - I'd prefer to see safer shebangs in scripts (`#!/usr/bin/env <>`), but this can be addressed later

**Pull request merged**